src/Microsoft.Restier.Core/ApiBuilderExtensions.cs

```diff
@@ -0,0 +1,564 @@
+        /// Call this to build a service chain explicitly.
+        /// Typically you just resolve the service with <see cref="IServiceProvider.GetService(Type)"/>, but
+        /// in case you register your own factory of <typeparamref name="T"/>, you may need this to get the
+        /// service chain registered with <see cref="ApiBuilder"/>.
+        /// </summary>
+        /// <typeparam name="T">The service type.</typeparam>
+        /// <param name="obj">The <see cref="ApiBuilder"/>.</param>
+        /// <returns>The service instance built.</returns>
+        /// <example>
+        /// <code>
+        /// services.AddScoped &lt; ISomeService &gt;(sp =>
+        ///     new SomeService(sp.BuildApiServiceChain &lt; ISomeService &gt;()));
+        /// </code>
+        /// </example>
+        public static T BuildApiServiceChain<T>(this IServiceProvider obj) where T : class
+        {
+            Ensure.NotNull(obj, nameof(obj));
+            return ChainedService<T>.DefaultFactory(obj);
+        }
+
+        private class SharedApiScopeFactory : IApiScopeFactory
+        {
+            public static readonly Func<IServiceProvider, IApiScopeFactory> Creator =
+                serviceProvider => new SharedApiScopeFactory(serviceProvider);
+
+            public SharedApiScopeFactory(IServiceProvider serviceProvider)
+            {
+                this.ServiceProvider = serviceProvider;
+            }
+
+            public IServiceProvider ServiceProvider
+            {
+                get; private set;
+            }
+
+            public IServiceScope CreateApiScope()
+            {
+                return new ServiceScope()
+                {
+                    ServiceProvider = this.ServiceProvider,
+                };
+            }
+
+            private class ServiceScope : IServiceScope
+            {
+                public IServiceProvider ServiceProvider
+                {
+                    get; set;
+                }
+
+                public void Dispose()
+                {
+                    this.ServiceProvider = null;
+                }
+            }
+        }
+
+        private class ContextApiScopeFactory : IApiScopeFactory
+        {
+            public static readonly Func<IServiceProvider, IApiScopeFactory> Creator =
+                serviceProvider => new ContextApiScopeFactory(
+                    serviceProvider.GetRequiredService<IServiceScopeFactory>());
+
+            public ContextApiScopeFactory(IServiceScopeFactory factory)
+            {
+                this.Factory = factory;
+            }
+
+            public IServiceScopeFactory Factory
+            {
+                get; private set;
+            }
+
+            public IServiceScope CreateApiScope()
+            {
+                return this.Factory.CreateScope();
+            }
+        }
+    }
+
+    internal static class ChainedService
+    {
+        public static Tuple<object, ConstructorInfo> CreateViaReflection(
```

Comment:
Design discussion: This part of logic duplicates with the functionality of DI. How about the following policy?

 - ctor. We don't care about how the ctor of a hook handler is implemented. We just let DI manage the instantiation itself.
 - public property `InnerHandler`. We ask the user to write such a public property in his hook handler class if the hook handler is supposed to be chained. Ignore the property if there is no such need. Then we use the similar logic in `BuildLegacyHandlers` to create the hook handler instances.